File: cpp-pthread/exer01a-max-div.cpp
/*
MAXIMUM NUMBER OF DIVISORS
*/


#include <iostream>
#include "../cpp-std/mylib-time.hpp"
using namespace std;



int main() {
    constexpr int RANGE_START = 1;
    constexpr int RANGE_END = 100000;

    int resValue = 0;
    int resNumDiv = 0;  // number of divisors of result

    auto tpStart = mylib::HiResClock::now();


    for (int i = RANGE_START; i <= RANGE_END; ++i) {
        int numDiv = 0;

        for (int j = i / 2; j > 0; --j)
            if (i % j == 0)
                ++numDiv;

        if (resNumDiv < numDiv) {
            resNumDiv = numDiv;
            resValue = i;
        }
    }


    auto timeElapsed = mylib::HiResClock::getTimeSpan(tpStart);

    cout << "The integer which has largest number of divisors is " << resValue << endl;
    cout << "The largest number of divisor is " << resNumDiv << endl;
    cout << "Time elapsed = " << timeElapsed.count() << endl;

    return 0;
}

File: js-nodejs/exer01a-max-div.js
/*
MAXIMUM NUMBER OF DIVISORS
*/


const hrtimeToNumber = hrtime => (hrtime[0] + (hrtime[1] / 1e9)).toFixed(6);


const mainFunc = () => {
  try {

    const RANGE_START = 1;
    const RANGE_END = 100000;

    let resValue = 0;
    let resNumDiv = 0;

    const tpStart = process.hrtime();

    for (let i = RANGE_START; i <= RANGE_END; ++i) {
      let numDiv = 0;

      for (let j = i / 2; j > 0; --j)
          if (i % j == 0)
              ++numDiv;

      if (resNumDiv < numDiv) {
          resNumDiv = numDiv;
          resValue = i;
      }
    }

    const timeElapsed = hrtimeToNumber(process.hrtime(tpStart));

    console.log('The integer which has largest number of divisors is', resValue);
    console.log('The largest number of divisor is', resNumDiv);
    console.log('Time elapsed =', timeElapsed);

  } catch (error) {
    console.error(error);
  }
};


mainFunc();
